installer/packer/scripts/harbor/configure_harbor.sh

```diff
@@ -0,0 +1,189 @@
+    sed -i -r "s%#?$cfg_key\s*=\s*.*%$cfg_key = $cfg_value%" $cfg
+  fi
+}
+
+function format {
+  file=$1
+  head=$(sed -rn 's/(-+[A-Za-z ]*-+)([^-]*)(-+[A-Za-z ]*-+)/\1/p' $file)
+  body=$(sed -rn 's/(-+[A-Za-z ]*-+)([^-]*)(-+[A-Za-z ]*-+)/\2/p' $file)
+  tail=$(sed -rn 's/(-+[A-Za-z ]*-+)([^-]*)(-+[A-Za-z ]*-+)/\3/p' $file)
+  echo $head > $file
+  echo $body | sed  's/\s\+/\n/g' >> $file
+  echo $tail >> $file
+}
+
+function genCert {
+  if [ ! -e $ca_cert ] || [ ! -e $ca_key ]
+  then
+    openssl req -newkey rsa:4096 -nodes -sha256 -keyout $ca_key \
+      -x509 -days 365 -out $ca_cert -subj \
+      "/C=US/ST=California/L=Palo Alto/O=VMware, Inc./OU=Containers on vSphere/CN=Self-signed by VMware, Inc."
+  fi
+  openssl req -newkey rsa:4096 -nodes -sha256 -keyout $key \
+    -out $csr -subj \
+    "/C=US/ST=California/L=Palo Alto/O=VMware/OU=Containers on vSphere/CN=$hostname"
+
+  echo "Add subjectAltName = IP: $ip_address to certificate"
+  echo subjectAltName = IP:$ip_address > $ext
+  openssl x509 -req -days 365 -in $csr -CA $ca_cert -CAkey $ca_key -CAcreateserial -extfile $ext -out $cert
+
+  echo "self-signed" > $flag
+  echo "Copy CA certificate to $ca_download_dir"
+  cp $ca_cert $ca_download_dir/
+}
+
+function secure {
+  ssl_cert=$(ovfenv -k harbor.ssl_cert)
+  ssl_cert_key=$(ovfenv -k harbor.ssl_cert_key)
+  if [ -n "$ssl_cert" ] && [ -n "$ssl_cert_key" ]; then
+    echo "ssl_cert and ssl_cert_key are both set, using customized certificate"
+    echo $ssl_cert > $cert
+    format $cert
+    echo $ssl_cert_key > $key
+    format $key
+    echo "customized" > $flag
+    return
+  fi
+
+  if [ ! -e $ca_cert ] || [ ! -e $cert ] || [ ! -e $key ]; then
+    echo "CA, Certificate or key file does not exist, will generate a self-signed certificate"
+    genCert
+    return
+  fi
+
+  if [ ! -e $flag ]; then
+    echo "The file which records the way generating certificate does not exist, will generate a new self-signed certificate"
+    genCert
+    return
+  fi
+
+  if [ ! $(cat $flag) = "self-signed" ]; then
+    echo "The way generating certificate changed, will generate a new self-signed certificate"
+    genCert
+    return
+  fi
+
+  cn=$(openssl x509 -noout -subject -in $cert | sed -n '/^subject/s/^.*CN=//p') || true
+  if [ "$hostname" !=  "$cn" ]; then
+    echo "Common name changed: $cn -> $hostname , will generate a new self-signed certificate"
+    genCert
+    return
+  fi
+
+  ip_in_cert=$(openssl x509 -noout -text -in $cert | sed -n '/IP Address:/s/.*IP Address://p') || true
+  if [ "$ip_address" !=  "$ip_in_cert" ]; then
+    echo "IP changed: $ip_in_cert -> $ip_address , will generate a new self-signed certificate"
+    genCert
+    return
+  fi
+
+  echo "Use the existing CA, certificate and key file"
+  echo "Copy CA certificate to $ca_download_dir"
+  cp $ca_cert $ca_download_dir/
+}
+
+function detectHostname {
+  hostname=$(hostnamectl status --static) || true
```

Comment:
`hostnamectl` might do this - the hostname you want for the cert is the one that dns provides, not the one configured for the system (yes, they could be the same). I've ended up using `nslookup` or equivalent once the appropriate interface is determined.
If there's multiple interfaces then this could be multi-valued and the alternates need to go into subjectAltName fields for IP address and DNS name respectively.